tests/cleanup_and_retention_test.go

```diff
@@ -1,191 +1,154 @@
+
+	retentionDays := 13
+	rowsToDelete := 10000
+	rowsToSpare := 1000
+	start := time.Now().AddDate(0, 0, -retentionDays).Add(-1 * time.Millisecond * time.Duration(rowsToDelete))
+
+	var binaryColumn string
+	switch db.DriverName() {
+	case "mysql":
+		binaryColumn = `binary(20)`
+	case "postgres":
+		binaryColumn = `bytea20`
 	}
 
 	t.Run("History", func(t *testing.T) {
-		db := database(t)
-		retentionDays := 1
-		scenario := scenarios[len(scenarios)-1]
-		wg := sync.WaitGroup{}
-
-		for _, stmt := range history.RetentionStatements {
-			_, err := db.Exec(fmt.Sprintf(
-				`CREATE TABLE %[1]s (%[2]s int NOT NULL, %[3]s bigint NOT NULL, PRIMARY KEY (%[2]s))`,
-				stmt.Table, stmt.PK, stmt.Column))
-			require.NoError(t, err)
-
-			_, err = db.Exec(fmt.Sprintf(`CREATE INDEX idx_%[1]s ON %[1]s (%[2]s)`, stmt.Table, stmt.Column))
-			require.NoError(t, err)
+		for _, stmt := range RetentionStatements {
+			withDisabledForeignKeyChecks(db, func() {
+				_, err := db.Exec(fmt.Sprintf(`DROP TABLE %s CASCADE`, stmt.Table))
+				require.NoError(t, err)
+				_, err = db.Exec(fmt.Sprintf(
+					`CREATE TABLE %[1]s (%[2]s %[4]s NOT NULL, %[3]s bigint NOT NULL, PRIMARY KEY (%[2]s))`,
+					stmt.Table, stmt.PK, stmt.Column, binaryColumn))
+				require.NoError(t, err)
+
+				_, err = db.Exec(fmt.Sprintf(`CREATE INDEX idx_%[1]s ON %[1]s (%[2]s)`, stmt.Table, stmt.Column))
+				require.NoError(t, err)
+			})
 
-			total := scenario.rowsToDelete + scenario.rowsToSpare
 			type row struct {
-				Id   int64
+				Id   []byte
 				Time int64
 			}
 			values := make([]row, 0, MaxPlaceholders)
-			start := time.Now().AddDate(0, 0, -retentionDays).Add(-1 * time.Millisecond * time.Duration(scenario.rowsToDelete))
 			startMilli := start.UnixMilli()
-			for j := int64(0); j < total; j++ {
-				if j == scenario.rowsToDelete {
+			for j := 0; j < (rowsToDelete + rowsToSpare); j++ {
+				if j == rowsToDelete {
 					startMilli += int64(2 * time.Minute)
 				}
 
-				values = append(values, row{j, startMilli + j})
+				id := make([]byte, 20)
+				binary.LittleEndian.PutUint64(id, uint64(j))
 
-				if len(values) == MaxPlaceholders || j == total-1 {
+				values = append(values, row{id[:], startMilli + int64(j)})
+
+				if len(values) == MaxPlaceholders || j == (rowsToDelete+rowsToSpare-1) {
 					_, err := db.NamedExec(fmt.Sprintf(`INSERT INTO %s (%s, %s) VALUES (:id, :time)`, stmt.Table, stmt.PK, stmt.Column), values)
 					require.NoError(t, err)
 					values = values[:0]
 				}
 			}
-
-			wg.Add(1)
 		}
 
-		ctx, cancelCtx := context.WithCancel(context.Background())
-		type result struct {
-			table   string
-			cleanup icingadb.CleanupResult
+		r := it.RedisServerT(t)
+		i := it.Icinga2NodeT(t, "master")
+		i.EnableIcingaDb(r)
+		i.Reload()
+		waitForDumpDoneSignal(t, r, 20*time.Second, 100*time.Millisecond)
+		it.IcingaDbInstanceT(t, r, rdb, services.WithIcingaDbConfig(fmt.Sprintf(`
+history-retention:
+  days: %d
+`, retentionDays)))
```

Comment:
Would be nice to specify explicit retention policies for two types, one just using a different value, and disabling cleanup for another one.

Reply:
Done.